### PR TITLE
feat(paddle): Add broadcast_shape function to Paddle Frontend

### LINKED ISSUE #23555: broadcast_shape

### PR DESCRIPTION
**Pull request by maxxies**
# PR Description

## Related Issue

Close #23555 

## Checklist

- [x] Did you add a function?
- [x] Did you add the tests?
- [x] Did you follow the steps we provided?

### CODE REVIEW

File: ivy/functional/frontends/paddle/math.py [hunk 1]
```diff
@@ -110,6 +110,14 @@ def atanh(x, name=None):
     return ivy.atanh(x)
 
 
+@with_supported_dtypes(
+    {"2.5.1 and below": ("float32", "float64", "int32", "int64")}, "paddle"
+)
+@to_ivy_arrays_and_back
+def broadcast_shape(x):
```
Comment (juliagsy) [inline ivy/functional/frontends/paddle/math.py:117]: Hey! not sure if I missed anything, but according to the docs there should be 2 parameters: https://www.paddlepaddle.org.cn/documentation/docs/en/api/paddle/broadcast_shape_en.html#broadcast-shape
Thanks!

Comment (maxxies) [reply ivy/functional/frontends/paddle/math.py:117]: Ok, sure. I would fix that

File: ivy_tests/test_ivy/test_frontends/test_paddle/test_math.py [hunk 2]
```diff
@@ -473,6 +473,35 @@ def test_paddle_atanh(
     )
 
 
+# broadcast_shape
+@handle_frontend_test(
+    fn_tree="paddle.broadcast_shape",
+    dtype_and_x=helpers.dtype_and_values(
+        available_dtypes=helpers.get_dtypes("valid"),
```
Comment (juliagsy) [inline ivy_tests/test_ivy/test_frontends/test_paddle/test_math.py:480]: please add a `num_arrays=2` configuration here for generating 2 arrays else the tests are raising an index out of range error when accessing `y=x[1]` thanks!

**Comment by maxxies** [comment]
@juliagsy can you have a quick review on the changes made? I am about to be removed from the application process due to slow progress.

**Comment by maxxies** [comment]
Hi,

Can I have a quick review on this pull request immediately, [unifyai/ivy]
feat(paddle): Add broadcast_shape function to Paddle Frontend (PR #23574).

My application process is about to be terminated.

Thanks.

File: ivy_tests/test_ivy/test_frontends/test_paddle/test_math.py [hunk 3]
```diff
@@ -473,6 +473,38 @@ def test_paddle_atanh(
     )
 
 
+# broadcast_shape
+@handle_frontend_test(
+    fn_tree="paddle.broadcast_shape",
+    dtype_and_x=helpers.dtype_and_values(
```
Comment (juliagsy) [inline ivy_tests/test_ivy/test_frontends/test_paddle/test_math.py:479]: Hey! Thanks for the changes! I've just come to realised the inputs to `broadcast_shape` expects shape integer or tuples, so I think this helper function is not the appropriate one, `helpers.get_shape` should be the one you're looking for. If necessary, you may generate data to 2 variables in the `handle_frontend_test` and pass them to the testing function.

Comment (maxxies) [reply ivy_tests/test_ivy/test_frontends/test_paddle/test_math.py:479]: Sure, let me fix that now

File: ivy/functional/frontends/paddle/math.py [hunk 4]
```diff
@@ -118,6 +118,12 @@ def atanh(x, name=None):
     return ivy.atanh(x)
 
 
+@with_supported_dtypes({"2.5.1 and below": ("int32", "int64")}, "paddle")
+@to_ivy_arrays_and_back
+def broadcast_shape(x, y):
+    return ivy.broadcast_shapes(*[x + y])
```
Comment (juliagsy) [inline ivy/functional/frontends/paddle/math.py:124]: I think you don't need to do this as the function signature in `ivy.broadcast_shapes` itself has `*` already, you may just pass in `...(x, y)`
Thanks!

Comment (maxxies) [reply ivy/functional/frontends/paddle/math.py:124]: ok

**Comment by maxxies** [comment]
@juliagsy can you review the updates?

File: ivy_tests/test_ivy/test_frontends/test_paddle/test_math.py [hunk 5]
```diff
@@ -473,6 +473,40 @@ def test_paddle_atanh(
...
+@handle_frontend_test(
+    fn_tree="paddle.broadcast_shape",
+    input_shape_x=helpers.get_shape(
+        min_num_dims=1,
+    ),
+    input_shape_y=helpers.get_shape(
+        min_num_dims=1,
+    ),
+)
+def test_paddle_broadcast_shape(
```
Comment (juliagsy) [inline ivy_tests/test_ivy/test_frontends/test_paddle/test_math.py:486]: Hey! Took some time to run the tests, thanks for the updates again! The tests seem to still be failing, could you please set up the environment and the tests can then be run through the command `python3 -m pytest ivy_tests/test_ivy/test_frontends/test_paddle/test_math.py::test_paddle_broadcast_shape`, and please resolve all tests failures after that, thanks a lot!
PS: if the generated shapes dimension is too large, feel free to limit it to a reasonable threshold, or you may also reference `test_broadcast_shapes`

Comment (maxxies) [reply ivy_tests/test_ivy/test_frontends/test_paddle/test_math.py:486]: Sure

Comment (maxxies) [reply ivy_tests/test_ivy/test_frontends/test_paddle/test_math.py:486]: I am sorry I had to wait on you every time to know what's happening with tests. I can't, at the moment, finish the setup to run the test individually. Is there a different approach I could get the results from the tests? plus what's the error that is being returned? Thanks @juliagsy

Comment (juliagsy) [reply ivy_tests/test_ivy/test_frontends/test_paddle/test_math.py:486]: I don't think you can get the results without setting up an environment unfortunately. There are several ways to setup an environment such as virtual env, conda, or docker (prob the preferred one to reduce conflicts), or codespaces (again, another preferred one as it's something like a vm and gives a clean env), hope this helps in setting up!
ref: https://unify.ai/docs/ivy/overview/contributing/setting_up.html

**Comment by charlie572** [comment]
This is a duplicate of #22977.

**Comment by maxxies** [comment]
> This is a duplicate of #22977.

Really, before I stated with this task, it had not been converted to an issue in the Add mathematical functions to Paddle Frontend issue list. So I converted it to an issue as stated in the Ivy docs. What do you reckon I do? @charlie572 @juliagsy

**Review by KareemMAX — CHANGES_REQUESTED** [review]
Hi, seems like you have changed `docs/demos` which you might have done by mistake. Can you revert that change because we can't merge with that change in place?

**Comment by charlie572** [comment]
> > This is a duplicate of #22977.
> 
> Really, before I stated with this task, it had not been converted to an issue in the Add mathematical functions to Paddle Frontend issue list. So I converted it to an issue as stated in the Ivy docs. What do you reckon I do? @charlie572 @juliagsy

I don't think the scripts were working when I created my issue, so it never updated the checklist. The scripts seem to work now.

**Comment by juliagsy** [comment]
@charlie572 thanks for flagging! seems like a glitch in the automation because the linked issue in the todo is the one from @maxxies 
As it's an issue on our side, we'll try to accept both PRs and incorporate both changes, the PRs may be proceed as usual :)

File: ivy/functional/frontends/paddle/math.py [hunk 6]
```diff
@@ -118,6 +118,12 @@ def atanh(x, name=None):
     return ivy.atanh(x)
 
 
+@with_supported_dtypes({"2.5.1 and below": ("int32", "int64")}, "paddle")
+@to_ivy_arrays_and_back
+def broadcast_shape(x, y):
```
Comment (juliagsy) [inline ivy/functional/frontends/paddle/math.py:123]: The args should be `x_shapes` and `y_shapes` according to the docs: https://www.paddlepaddle.org.cn/documentation/docs/en/2.2/api/paddle/broadcast_shape_en.html

File: ivy/functional/frontends/paddle/math.py [hunk 7]
```diff
@@ -118,6 +118,12 @@ def atanh(x, name=None):
...
+    return ivy.broadcast_shapes(x, y)
```
Comment (juliagsy) [inline ivy/functional/frontends/paddle/math.py:124]: this should be updated accordingly

File: ivy_tests/test_ivy/test_frontends/test_paddle/test_math.py [hunk 8]
```diff
@@ -473,6 +473,35 @@ def test_paddle_atanh(
     )
 
 
+# broadcast_shape
+@handle_frontend_test(
+    fn_tree="paddle.broadcast_shape",
+    input_shapes_x=helpers.mutually_broadcastable_shapes(
```
Comment (juliagsy) [inline ivy_tests/test_ivy/test_frontends/test_paddle/test_math.py:479]: you can try importing: `import hypothesis.extra.numpy as nph`
and use:
```
input_shapes_x=nph.mutually_broadcastable_shapes(
        num_shapes=2, min_dims=1, max_dims=5, min_side=1, max_side=5
    ),
```

File: ivy_tests/test_ivy/test_frontends/test_paddle/test_math.py [hunk 9]
```diff
@@ -473,6 +473,35 @@ def test_paddle_atanh(
...
+def test_paddle_broadcast_shape(
+    *,
+    input_shapes_x,
+    on_device,
+    fn_tree,
+    frontend,
+    backend_fw,
+    test_flags,
+):
+    shapes = input_shapes_x
```
Comment (juliagsy) [inline ivy_tests/test_ivy/test_frontends/test_paddle/test_math.py:492]: this is unnecessary, you can use `input_shapes_x` directly.

File: ivy_tests/test_ivy/test_frontends/test_paddle/test_math.py [hunk 10]
```diff
@@ -473,6 +473,35 @@ def test_paddle_atanh(
...
+    helpers.test_frontend_function(
+        input_dtypes=["int32", "int64"],
+        frontend=frontend,
+        backend_to_test=backend_fw,
+        test_flags=test_flags,
+        fn_tree=fn_tree,
+        on_device=on_device,
+        x=shapes[0],
```
Comment (juliagsy) [inline ivy_tests/test_ivy/test_frontends/test_paddle/test_math.py:500]: this should be something like
```
        x_shape=input_shapes_x[0][0],
        y_shape=input_shapes_x[0][1],
```

**Review by juliagsy — CHANGES_REQUESTED** [review]
Hey1 The tests are still not passing unfortunately, I've double check the docs and apparently the variable names are diff, they have to be identical to native. Could you please try the below suggested change and continue working on getting the tests to pass? Thanks!

**Comment by maxxies** [comment]
> Hey1 The tests are still not passing unfortunately, I've double check the docs and apparently the variable names are diff, they have to be identical to native. Could you please try the below suggested change and continue working on getting the tests to pass? Thanks!

This is the error I still get:

 +---------------- 2 ----------------
    | Traceback (most recent call last):
    |   File "/workspaces/ivy/ivy_tests/test_ivy/test_frontends/test_paddle/test_math.py", line 493, in test_paddle_broadcast_shape
    |     helpers.test_frontend_function(
    |   File "/workspaces/ivy/ivy_tests/test_ivy/helpers/function_testing.py", line 981, in test_frontend_function
    |     value_test(
    |   File "/workspaces/ivy/ivy_tests/test_ivy/helpers/assertions.py", line 141, in value_test
    |     assert len(ret_np_flat) == len(ret_np_from_gt_flat), (
    | AssertionError: The length of results from backend paddle and ground truth framework paddle does not match
    | 
    | len(ret_np_flat) != len(ret_np_from_gt_flat):
    | 
    | ret_np_flat:
    | 
    | [array(1)]
    | 
    | ret_np_from_gt_flat:
    | 
    | []
    | Falsifying example: test_paddle_broadcast_shape(
    |     # The test always failed when commented parts were varied together.
    |     on_device='cpu',
    |     frontend='paddle',
    |     backend_fw='paddle',
    |     input_shapes_x=BroadcastableShapes(input_shapes=((1,), (1,)), result_shape=(1,)),  # or any other generated value
    |     fn_tree='ivy.functional.frontends.paddle.broadcast_shape',
    |     test_flags=FrontendFunctionTestFlags(
    |         num_positional_args=0,
    |         with_out=False,
    |         inplace=False,
    |         as_variable=[False],
    |         native_arrays=[False],
    |         test_compile=False,
    |         generate_frontend_arrays=False,
    |         transpile=False,
    |         precision_mode=False,
    |     ),  # or any other generated value
    | )
    | 
    | You can reproduce this example by temporarily adding @reproduce_failure('6.87.0', b'AXicY2BAAQAADwAB') as a decorator on your test case

**Comment by maxxies** [comment]
> ret_np_from_gt_flat

This `ret_np_from_gt_flat` returns an empty array

**Comment by juliagsy** [comment]
Could you please try adding `test_values=False`? then check the values manually? Might be because they are tuples or lists instead of arrays. Some references using `test_values=False`: test_jax_cholesky, test_jax_eig

**Comment by maxxies** [comment]
> Could you please try adding `test_values=False`? then check the values manually? Might be because they are tuples or lists instead of arrays. Some references using `test_values=False`: test_jax_cholesky, test_jax_eig

I think the tuples are handled appropriately but the assertions still fails:

`                kwargs["out"] = out
                if is_ret_tuple:
                    if test_flags.generate_frontend_arrays:
                        flatten_ret = flatten_frontend(
                            ret=ret,
                            backend=backend_to_test,
                            frontend_array_fn=create_frontend_array,
                        )
                        flatten_out = flatten_frontend(
                            ret=out,
                            backend=backend_to_test,
                            frontend_array_fn=create_frontend_array,
                        )
                    else:
                        flatten_ret = flatten(backend=backend_to_test, ret=ret)
                        flatten_out = flatten(backend=backend_to_test, ret=out)
                    for ret_array, out_array in zip(flatten_ret, flatten_out):
                        if ivy_backend.native_inplace_support and not any(
                            (ivy_backend.isscalar(ret), ivy_backend.isscalar(out))
                        ):
                            if test_flags.generate_frontend_arrays:
                                assert ret_array.ivy_array.data is out_array.ivy_array.data
                            else:
                                assert ret_array.data is out_array.data
>                       assert ret_array is out_array
E                       AssertionError
E                       Falsifying example: test_paddle_broadcast_shape(
E                           on_device='cpu',
E                           frontend='paddle',
E                           backend_fw='paddle',
E                           input_shapes_x=BroadcastableShapes(input_shapes=((1,), (1,)), result_shape=(1,)),  # or any other generated value
E                           test_flags=FrontendFunctionTestFlags(
E                               num_positional_args=0,
E                               with_out=True,
E                               inplace=False,
E                               as_variable=[False],
E                               native_arrays=[False],
E                               test_compile=False,
E                               generate_frontend_arrays=True,
E                               transpile=False,
E                               precision_mode=False,
E                           ),
E                           fn_tree='ivy.functional.frontends.paddle.broadcast_shape',
E                       )
E                       Explanation:
E                           These lines were always and only run by failing examples:
E                               /workspaces/ivy/ivy_tests/test_ivy/helpers/function_testing.py:772
E                               /workspaces/ivy/ivy_tests/test_ivy/helpers/function_testing.py:775
E                               /workspaces/ivy/ivy_tests/test_ivy/helpers/function_testing.py:778
E                               /workspaces/ivy/ivy_tests/test_ivy/helpers/function_testing.py:779
E                               /workspaces/ivy/ivy_tests/test_ivy/helpers/function_testing.py:809
E                               /workspaces/ivy/ivy_tests/test_ivy/helpers/function_testing.py:814
E                       
E                       You can reproduce this example by temporarily adding @reproduce_failure('6.87.0', b'AXicY2CAAUYQAgAAFgAD') as a decorator on your test case

ivy_tests/test_ivy/helpers/function_testing.py:830: AssertionError`

P.S:

It passes for Jax

**Comment by juliagsy** [comment]
Hey! Apologies for the delay again! I've been looking into the code and the related function to see where I'm able to help but haven't seen something too off yet. I'll also cc some engineers from the frontends/testing team to see if they have an idea on the issue about gt returning empty list/results. (cc @AnnaTz @sherry30) thanks all!

**Comment by maxxies** [comment]
> Hey! Apologies for the delay again! I've been looking into the code and the related function to see where I'm able to help but haven't seen something too off yet. I'll also cc some engineers from the frontends/testing team to see if they have an idea on the issue about gt returning empty list/results. (cc @AnnaTz @sherry30) thanks all!

Sure, thanks.

**Review by juliagsy — COMMENTED** [review]
Hey! Just a quick one, could you quickly go into `docs/demos` and do `git checkout aaef6cb` then `cd` back to ivy root and do `git add docs/demos` then commit to your own branch? There's a diff showing in `docs/demos` in the PR and we don't want to commit any changes in there yet, thanks!
